parser_api/example.rb

```diff
@@ -0,0 +1,173 @@
+require 'puppet'
+
+# Example of a module setting everything up to perform custom
+# validation of an AST model produced by parsing puppet source.
+#
+module MyValidation
+
+  # A module for the new issues that the this new kind of validation will generate
+  #
+  module Issues
+    # (see Puppet::Pops::Issues#issue)
+    # This is boiler plate code
+    def self.issue (issue_code, *args, &block)
+      Puppet::Pops::Issues.issue(issue_code, *args, &block)
+    end
+
+    INVALID_WORD = issue :INVALID_WORD, :text do
+      "The word '#{text}' is not a real word."
+    end
+  end
+
+  # This is the class that performs the actual validation by checking input
+  # and sending issues to an acceptor.
+  #
+  class MyChecker
+    attr_reader :acceptor
+    def initialize(diagnostics_producer)
+      @@bad_word_visitor       ||= Puppet::Pops::Visitor.new(nil, "badword", 0, 0)
```

Comment:
The concept of the visitor/polymorphic dispatcher must be explained. To go from this line what actually happens (calls to `badword_QualifiedName` etc.) is not self evident.